### PR TITLE
Refactor QNAP and airOS modules for creds

### PR DESCRIPTION
**Pull request by wvu**
- [x] Review `auxiliary/gather/qnap_backtrace_admin_hash`
- [x] Review `exploit/linux/ssh/ubiquiti_airos_file_upload`

### CODE REVIEW

File: modules/exploits/linux/ssh/ubiquiti_airos_file_upload.rb [hunk 1]
```diff
@@ -148,11 +148,10 @@ def ssh_login
         refs: self.references,
         info: ssh.transport.server_version.version
       )
-      report_note(
-        host: rhost,
-        port: datastore['SSH_PORT'],
-        type: 'airos.ssh.key',
-        data: private_key
+      store_valid_credential(
```
Comment (wvu-r7) [inline modules/exploits/linux/ssh/ubiquiti_airos_file_upload.rb]: I skipped the password here, since we really only care about the key we used?

Comment (wvu-r7) [inline modules/exploits/linux/ssh/ubiquiti_airos_file_upload.rb]: I skipped the password here, since we really only care about the key we used?

Comment (wvu-r7) [inline modules/exploits/linux/ssh/ubiquiti_airos_file_upload.rb]: I skipped the password here, since we really only care about the key we used...

Comment (jmartin-r7) [inline modules/exploits/linux/ssh/ubiquiti_airos_file_upload.rb]: In this case the module uses 2 library mixins, we don't yet provide helper service_details method on ssh and that looks to be the service you want this credential on.  May be worth adding an override for service_details port and protocol value here.

```
def service_details
  super.merge({ 
    service_name: 'ssh',
    port: datastore['SSH_PORT']
  })
end
```

Comment (wvu-r7) [inline modules/exploits/linux/ssh/ubiquiti_airos_file_upload.rb]: Oops, overlooked that detail while reading the method. Thanks!

File: modules/auxiliary/gather/qnap_backtrace_admin_hash.rb [hunk 2]
```diff
@@ -90,11 +90,10 @@ def run
 
     if admin_hash
       print_good("Hopefully this is your hash: #{admin_hash}")
-      report_note(
-        host: rhost,
-        port: rport,
-        type: 'qnap.admin.hash',
-        data: admin_hash
+      store_valid_credential(
```
Comment (wvu-r7) [inline modules/auxiliary/gather/qnap_backtrace_admin_hash.rb]: @jmartin-r7: So what I'm worried about here is that `service_details` should be set from `HttpClient`, which means `store_valid_credential` should be creating a login object, but this is a non-replayable hash that - if I remember the creds API correctly - should not have a login object associated with it. Can you clarify this for me?

Comment (dmaloney-r7) [inline modules/auxiliary/gather/qnap_backtrace_admin_hash.rb]: A login indicates the credential object can actually log in to the specified service. But service information can also be used in the origin which says where the credential came from. It looks like this new method that has been added does not account for a situation where you don't want a login to be created but it's a service origin. I would recommend actually using the create_credential method instead here

Comment (jmartin-r7) [inline modules/auxiliary/gather/qnap_backtrace_admin_hash.rb]: store_vaild_credential is expected to cover the 80% case for modules, service only creds look to have different expectations. Maybe something like

```
    connection_details = {
          module_fullname: self.fullname,
          username: 'admin',
          private: admin_hash,
          private_type: :nonreplayable_hash,
          workspace_id: myworkspace_id,
      }.merge(service_details)
      create_credential(connection_details)
```

I doesn't save as much code however if does take advantage of the standard service details.

Comment (wvu-r7) [inline modules/auxiliary/gather/qnap_backtrace_admin_hash.rb]: I guess there was no escaping it after all.

**Comment by jmartin-r7** [comment]
# Release Note

Improve credential reporting in auxiliary/gather/qnap_backtrace_admin_hash & exploit/linux/ssh/ubiquiti_airos_file_upload

**Comment by wvu-r7** [comment]
You rock, @jmartin-r7!

**Pull request merged**